### PR TITLE
🐕‍🦺 Feat/6 API Service Layer

### PR DESCRIPTION
**Pull request by demariadaniel**
https://github.com/Pan-Canadian-Genome-Library/daco/issues/6

- Adds Application Service w/ create, find by id, list many, update, delete functionality
  - this template can be adapted to other Models
- Includes filtering / sorting as described in ticket
- Full set of unit tests
- Separates implementations for connecting to live DB / test-DB
- Add `updatedAt` field to Applications
- Fix a few relation fields

### CODE REVIEW

File: apps/api/.env.test
Comment (demariadaniel) [inline apps/api/.env.test:8]: Just differentiating from .env.schema values

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [inline apps/api/src/service/application-service.ts:60]: Not listed in the ticket but created out of the need to update records in the tests

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [inline apps/api/src/service/application-service.ts:150]: Similar to `findOneAndUpdate`, there was a need to delete records after they were created.

There were connection issues (now fixed) creating two separate DB connections (one for the service, one for the tests) so I needed additional service methods first to resolve the test issues before I diagnosed the connection issues.

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [inline apps/api/src/service/application-service.ts:26]: Open to alternative approaches to passing the database connection reference as an argument

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [inline apps/api/src/tests/services/application.ts:105]: Better variable names?

File: apps/api/src/main.ts
Comment (demariadaniel) [inline apps/api/src/main.ts:23]: Adding separation between starting server and starting test-server

File: apps/api/src/tests/db/index.ts
Comment (demariadaniel) [inline apps/api/src/tests/db/index.ts:47]: These are tests from the previous PR which do not have a Service yet (test plain Drizzle actions)

**Review by JamesTLopez — COMMENTED** [review]
Just some debugging console.logs and perhaps naming improvments with the docker file

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts:51]: Yeah fair point, this loop is definitely just a placeholder for a batch insert, or populating the test DB on init. Could be improved for sure, added that pretty quick at the end of the day yesterday.

However, the service level logs were intentional; there is a decent amount of logging in our Argo Clinical Service indicating which endpoints were called with what arguments.
So I think a base level of reporting is beneficial too.

examples:
Screenshot 2024-11-15 at 2 29 36 PM

Screenshot 2024-11-15 at 2 24 59 PM

File: apps/api/docker-compose.yml
Comment (demariadaniel) [reply apps/api/docker-compose.yml:4]: Good call, added

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts]: Commented on: https://github.com/orgs/Pan-Canadian-Genome-Library/projects/1/views/1?pane=issue&itemId=87273850&issue=Pan-Canadian-Genome-Library%7Cdaco%7C44

We will standardize logging when we get to the REST side of things

File: apps/api/src/service/application-service.ts
Comment (joneubank) [inline apps/api/src/service/application-service.ts:107]: This encodes a lot of rules for the content of the "sort" string value, and our type system won't validate if they are formatted correctly or contain valid fields for our data model. If we do use a string like this, we should add TSDoc comments on this function to ensure the content of this parameter are well documented. I have an alternate suggestion though:

I see the `-columnName` pattern for sort parameters used when encoded in URL query params, where you want to put condensed information and only string values. We don't have these restriction on our internal functions parameters. We can explicitly require our sort paramater to have two properties, one for direction and the other for the column name, and the values for these properties can be strictly set as `asc | desc` and as the list of legal column names for the schema you are sorting.

It would look something like:

```ts
type OrderBy<T extends string> = {
	direction: 'asc' | 'desc';
	column: T;
};
type ApplicationsColumnName = keyof typeof applications.$inferSelect;

// the sort function argument can be typed as:
(sort?: Array<OrderBy<ApplicationsColumnName>>) => { 

// Making this an array allows the function to orderBy multiple properties, sorting by user ascending and then created_at descending, for instance.

// The orderBy clause can be built like:
const orderByArguments = sort
	? sort.map((sortBy) =>
			sortBy.direction === 'asc' ? asc(applications[sortBy.column]) : desc(applications[sortBy.column]),
	  )
	: [asc(applications.created_at)];

	await db.select({}).orderBy(...orderByArguments);
}
```

Lastly, if we expect to do db sorting in multiple places, then we could build some of this code as common utilities, for sort types and to built the orderby arguments.

File: apps/api/src/service/application-service.ts
Comment (joneubank) [reply apps/api/src/service/application-service.ts:154]: We have no use case for deleting applications. In order to prevent this being used when we shouldnt, can we remove this function?

File: apps/api/src/tests/services/application.ts
Comment (joneubank) [inline apps/api/src/tests/services/application.ts:81]: The next set of tests are for the listApplications, can we group them into a `suite('listApplications', async ()=>{...})`?

File: apps/api/src/tests/services/application.ts
Comment (joneubank) [inline apps/api/src/tests/services/application.ts:96]: How do you initialize the data for this test? It looks like it depends on previous testa and that if I run this test directly it will fail. We want our unit tests to be repeatable and standalone, able to run in any order or even on their own.

If there is a series of steps that we can't separate to test independently, and would be valuable to test in sequence, then we should build these as a single test with sub tests. That way we are assured to run them in sequence.

Comment (joneubank) [reply apps/api/src/tests/services/application.ts:96]: This comment applies to many tests in this file. We should have an script that inserts a test data set before running the tests that read data. We can use our create application api to insert these records.

File: apps/api/src/tests/services/application.ts
Comment (joneubank) [inline apps/api/src/tests/services/application.ts:95]: It may make more sense to ensure that every record in the response has the correct state, instead of just checking the count... unless you have cleared the db and initialized it with a precise test data set.

File: apps/api/src/tests/services/application.ts
Comment (joneubank) [reply apps/api/src/tests/services/application.ts:106]: This is fine really.

File: apps/api/src/tests/services/application.ts
Comment (joneubank) [inline apps/api/src/tests/services/application.ts:122]: TODO: Tests need to be independent or grouped together as a single test with sub tests.

**Review by joneubank — CHANGES_REQUESTED** [review]
Solid foundation here, lots of smart additions. Thanks!

File: apps/api/.env.test
Comment (joneubank) [reply apps/api/.env.test:8]: On second review, that example may not work with our tests, since the env might be set by the `before()` function after the env is read by the startDb function... Need to test and confirm, I have ideas if need be. I stand by the idea that we shouldn't be feeding an env file into our test suite except to configure/control what tests are running or how they are being run.

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts:122]: Yes definitely. I parked some of the test setup tasks because Drizzle seeding is not available yet (see comment at line 56 of this file).

The tests are written sequentially (so that Test B depends on Test A etc.) but this is not ideal (or sustainable), and I intended to update this before we went much further.

I opened the PR at this state because it was functionally complete, and I wanted to take a step back to assess how we should approach test setup with the team's input.

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts:96]: See below comment -- I'm aware the test setup could use improvements and felt that was one piece worth revisiting. I should have expanded on my notes so you were aware too!

File: apps/api/drizzle/0000_nosy_echo.sql
Comment (demariadaniel) [reply apps/api/drizzle/0000_nosy_echo.sql:75]: Yes agreed, I was thinking we'd be adding migration files as soon as there's anything deployed. Very good point about informing rest of team re: running migrations.

File: apps/api/drizzle/0000_nosy_echo.sql
Comment (leoraba) [inline apps/api/drizzle/0000_nosy_echo.sql:1]: I think you can generate migration files and give a more meaningful name using `npx drizzle-kit generate --name=init` 
doc:  https://orm.drizzle.team/docs/drizzle-kit-generate#custom-migration-file-name

Comment (leoraba) [reply apps/api/drizzle/0000_nosy_echo.sql:1]: or just rename it manually in 😜

File: apps/api/src/db/schemas/applications.ts
Comment (leoraba) [inline apps/api/src/db/schemas/applications.ts:1]: This schema change is not reflected in the docs. May I suggest a tool to generate .dbml files e.g. [drizzle-dbml-generator](https://www.npmjs.com/package/drizzle-dbml-generator)
That task can be added in package.json on build time for example to generate .dbml and keep the docs updated

**Review by leoraba — COMMENTED** [review]
LGTM. just left some comments 👍

File: apps/api/src/db/index.ts
Comment (demariadaniel) [reply apps/api/src/db/index.ts]: All good, improved to use connectToDb, important distinction

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts]: Removed

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts:82]: Yes, added a `describe()` which is just alias for suite, later we can add same for create/get etc.

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts]: Yep, I've grouped these in a transaction

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts:65]: No, so I've added one to start the FindOneAndUpdate suite

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts]: Update/set types are not well documented but finally figured this out

File: apps/api/src/service/application-service.ts
Comment (demariadaniel) [reply apps/api/src/service/application-service.ts]: Also added types/utils files

File: apps/api/.env.test
Comment (demariadaniel) [reply apps/api/.env.test]: Setting process.env does work but it's not actually necessary because we can just set the container variables and obtain the connection string like so:

```
container = await new PostgreSqlContainer()
			.withUsername(PG_USER)
			.withPassword(PG_PASSWORD)
			.withDatabase(PG_DATABASE)
			.start();

const connectionString = container.getConnectionUri();
db = connectToDb(connectionString);
```

where PG_USER etc. are defined as constants in a test file

File: apps/api/src/db/index.ts
Comment (demariadaniel) [reply apps/api/src/db/index.ts]: Moved to testUtils

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts]: Updated to confirm state values, as well above user_id test now checks user_id

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts]: Added test-data init functions to testUtils

File: apps/api/src/tests/services/application.ts
Comment (demariadaniel) [reply apps/api/src/tests/services/application.ts]: Tests are now grouped together and setup functions enable independent execution

File: apps/api/src/db/schemas/applications.ts
Comment (demariadaniel) [reply apps/api/src/db/schemas/applications.ts:1]: OK I added a script to generate the DBML, using schema generated by `drizzle-kit pull`
The drizzle-config is using a glob path string which doesn't work with pgGenerate
So rather than refactor all the schema / table / enum / relations code `pull` works

However, there is a conflict w/ `files.content` field which will need to be revisited
I've left a TODO comment and this can be reviewed when we do work on Files service

File: packages/data-model/src/types.ts
Comment (joneubank) [inline packages/data-model/src/types.ts:30]: Important to use these named enum values if we are using TS enums, so thank you for this update. We can take a moment to review our enum usage in a future change.

File: apps/api/drizzle/schema.ts
Comment (demariadaniel) [reply apps/api/drizzle/schema.ts:66]: This was generated using `drizzle-kit pull` so it's the auto-generated upper limit for a BigInt

File: apps/api/src/service/application-service.ts
Comment (leoraba) [inline apps/api/src/service/application-service.ts:36]: 👍

File: apps/api/drizzle/schema.ts
Comment (demariadaniel) [reply apps/api/drizzle/schema.ts:115]: `If character varying (or varchar) is used without length specifier, the type accepts strings of any length.`
https://www.postgresql.org/docs/current/datatype-character.html
https://orm.drizzle.team/docs/column-types/pg#varchar
Drizzle says nothing so referring to PG docs

File: apps/api/src/tests/index.ts
Comment (leoraba) [inline apps/api/src/tests/index.ts:1]: just wondering, should `test/` folder be at top level along the `src/`. also consider changing extension to differentiate the actual code like `.test.ts` or `.spec.ts`

Comment (demariadaniel) [reply apps/api/src/tests/index.ts:1]: True, argo-clinical does have `test` at the same level as `src` and we have generally used the `.spec` convention

I need these changes in `main` to unblock other work so I can incorporate these suggestions into the next branch

**Review by leoraba — APPROVED** [review]
LGTM 🎉

File: apps/api/drizzle/schema.ts
Comment (demariadaniel) [reply apps/api/drizzle/schema.ts:66]: Leo and I chatted about this -- there is a disconnect between the DBML schema input (expects an object) and the Drizzle config schema definition (string glob path to schemas folder './schemas/**')

This should be refactored so that both point to the same schema definition, my only hesitation is to limit the scope of this PR. I will update this and the test files in the next branch.

**Pull request merged**
